core/shared/src/main/scala/zio/Dequeue.scala

```diff
@@ -85,27 +85,20 @@ trait Dequeue[+A] extends Serializable {
    * maximum. If there are fewer than the minimum number of elements available,
    * suspends until at least the minimum number of elements have been collected.
    */
-  final def takeBetween(min: Int, max: Int)(implicit trace: Trace): UIO[Chunk[A]] =
-    ZIO.suspendSucceed {
-
-      def takeRemainder(min: Int, max: Int, acc: Chunk[A]): UIO[Chunk[A]] =
-        if (max < min) ZIO.succeed(acc)
-        else
-          takeUpTo(max).flatMap { bs =>
-            val remaining = min - bs.length
-            if (remaining == 1)
-              take.map(b => acc ++ bs :+ b)
-            else if (remaining > 1) {
-              take.flatMap { b =>
-                takeRemainder(remaining - 1, max - bs.length - 1, acc ++ bs :+ b)
-
-              }
-            } else
-              ZIO.succeed(acc ++ bs)
-          }
-
-      takeRemainder(min, max, Chunk.empty)
-    }
+  final def takeBetween(min: Int, max: Int)(implicit trace: Trace): UIO[Chunk[A]] = {
+    def takeRemainder(min: Int, max: Int, acc: Chunk[A]): UIO[Chunk[A]] =
+      if (max < min) ZIO.succeed(acc)
+      else
+        takeUpTo(max).flatMap { bs =>
+          val remaining = min - bs.length
+
+          if (remaining <= 0) Exit.succeed(acc ++ bs)
+          else if (remaining == 1) take.map(b => acc ++ bs :+ b)
+          else take.flatMap(b => takeRemainder(remaining - 1, max - bs.length - 1, acc ++ bs :+ b))
```

Comment:
@guizmaii I'm thinking that while this is okay for small `min` / `max` numbers, it's probably quite unoptimized for large numbers due to the concatenation of `Chunk`s. I wonder if we could use a ChunkBuilder instead when `min` or `max` are relatively large.

However perhaps lets look at this at another PR

Reply:
That's a good question. I don't really understand well how `Chunk`/`ChunkBuilder` are working nor how to get best performances from then, especially about concatenating two `Chunk`/`ChunkBuilder`

I know that Chunks concatenation doesn't relocate immediately a new Chunk containing both. It's using `final case class Concat[A](override protected val left: Chunk[A], override protected val right: Chunk[A])` to concat lazily

Why do you think using `ChunkBuilder` would help?